desc/objectives/objective_funs.py

```diff
@@ -272,23 +172,65 @@ def build(self, eq=None, use_jit=None, verbose=1):
         for objective in self.objectives:
             if verbose > 0:
                 print("Building objective: " + objective.name)
-            objective.build(eq, use_jit=self.use_jit, verbose=verbose)
+            objective.build(use_jit=self.use_jit, verbose=verbose)
             self._dim_f += objective.dim_f
         if self._dim_f == 1:
             self._scalar = True
         else:
             self._scalar = False
 
-        self.set_args()
         self._set_derivatives()
         if self.use_jit:
             self.jit()
 
+        self._set_things()
+
         self._built = True
         timer.stop("Objective build")
         if verbose > 1:
             timer.disp("Objective build")
 
+    def _set_things(self, things=None):
+        """Tell the ObjectiveFunction what things it is optimizing.
+
+        Parameters
+        ----------
+        things : list, tuple, or nested list, tuple of Optimizable
+            Collection of things used by this objective. Defaults to all things from
+            all sub-objectives.
+
+        Notes
+        -----
+        Sets ``self._flatten`` as a function to return unique flattened list of things
+        and ``self._unflatten`` to recreate full nested list of things
+        from unique flattened version.
+        """
+        from jax.tree_util import tree_flatten, tree_unflatten
+
+        things = setdefault(things, [obj.things for obj in self.objectives])
+
+        flat_, treedef_ = tree_flatten(
+            things, is_leaf=lambda x: isinstance(x, Optimizable)
+        )
+        unique_, inds_ = unique_list(flat_)
```

Comment:
question to make sure I understand: our classes are also JAX pytrees right?

so when `flat_` comes out, does it contain the attributes which make up each class, and then this line gets only the unique ones, so we know what the the unique list of parameters we need to recreate the `things` are?

or is `flat_` the underlying array values themselves?

In the[ JAX pytrees page](https://jax.readthedocs.io/en/latest/pytrees.html) it seems like flat would just return the array values, but then I am unsure of how this unique_list helps here

Reply:
Ah ok wait I see, it is actually the objects themselves

```python
Building objective: plasma-vessel distance
Precomputing transforms
treedef_=PyTreeDef([[*, *]])
before unique: flat_=[Equilibrium at 0x7fa93414a710 (L=24, M=12, N=4, NFP=12.0, sym=1, spectral_indexing=fringe), FourierRZToroidalSurface at 0x7fa934297190 (name=)]
after  unique: flat_=[Equilibrium at 0x7fa93414a710 (L=24, M=12, N=4, NFP=12.0, sym=1, spectral_indexing=fringe), FourierRZToroidalSurface at 0x7fa934297190 (name=)]
```

Reply:
yeah, the `is_leaf` function stops the flattening when it encounters an `Optimzable` object

Reply:
The `unique_list` part is to handle the possibility of the same object being in the list multiple times (ie, because the object is connected to multiple objectives)

Reply:
Got it, like a `PlasmaVesselDistance` and a QS touching the same `Equilibrium` or something